kt/plugins/godot-intellij-plugin/src/main/kotlin/godot/intellij/plugin/inspections/registration/inspector/RegisterPropertiesInspection.kt

```diff
@@ -0,0 +1,278 @@
+import godot.core.Color
+import godot.intellij.plugin.inspections.registration.quickfix.RegisterPropertyMutabilityQuickFix
+import org.jetbrains.kotlin.backend.common.serialization.findPackage
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.descriptors.impl.ClassConstructorDescriptorImpl
+import org.jetbrains.kotlin.descriptors.impl.SimpleFunctionDescriptorImpl
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.idea.inspections.AbstractKotlinInspection
+import org.jetbrains.kotlin.idea.intentions.loopToCallChain.isConstant
+import org.jetbrains.kotlin.idea.refactoring.fqName.fqName
+import org.jetbrains.kotlin.idea.util.findAnnotation
+import org.jetbrains.kotlin.js.descriptorUtils.getJetTypeFqName
+import org.jetbrains.kotlin.js.resolve.diagnostics.findPsi
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.nj2k.postProcessing.type
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.referenceExpression
+import org.jetbrains.kotlin.resolve.bindingContextUtil.getReferenceTargets
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.descriptorUtil.isCompanionObject
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.typeUtil.*
+
+class RegisterPropertiesInspection : AbstractKotlinInspection() {
+    private val mutabilityQuickFix by lazy { RegisterPropertyMutabilityQuickFix() }
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return propertyVisitor { ktProperty ->
+            if (ktProperty.findAnnotation(FqName(RegisterProperty::class.java.canonicalName)) != null) {
+                checkMutability(ktProperty, holder)
+                checkRegisteredType(ktProperty, holder)
+                checkIfDefaultValueIsConstant(ktProperty, holder)
+            }
+            checkPropertyHintAnnotations(ktProperty, holder) //outside to check if the property is also registered
+        }
+    }
+
+    private fun checkMutability(ktProperty: KtProperty, holder: ProblemsHolder) {
+        if (!ktProperty.isVar) {
+            holder.registerProblem(ktProperty.valOrVarKeyword, "Registered properties have to be mutable", mutabilityQuickFix)
+        }
+    }
+
+    private fun checkRegisteredType(ktProperty: KtProperty, holder: ProblemsHolder) {
+        val type = ktProperty.type() ?: return
+        if (type.getJetTypeFqName(false).startsWith("kotlin.collections") && ktProperty.findAnnotation(FqName("godot.annotation.EnumFlag")) == null) { //enum flag is the only case where registering a kotlin collection is allowed
+            holder.registerProblem(getInitializerProblemLocation(ktProperty), "Kotlin collections cannot be registered as default values. Consider using one of the godot collections. There are handy conversion functions available")
```

Comment:
Can't we add a quick fix using collection fqdn to provide the conversion ?

Reply:
Sure we can. But for this case it might not be that trivial. If it's ok for you I'll just add a todo for now and add them later once the whole plugin is a bit more mature

Reply:
I'm fine with it :-)